pkg/objects/github.go

```diff
@@ -0,0 +1,277 @@
+	s.httpClient = &http.Client{}
+
+	// Use a separate http.Client for creating github.Client.
+	client2 := &http.Client{}
+	if s.Token != "" {
+		src := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: string(s.Token)})
+		client2 = oauth2.NewClient(context.Background(), src)
+	}
+	if s.url.Hostname() != "github.com" {
+		baseURL := strings.TrimSuffix(s.url.String(), s.url.Path)
+		var err error
+		s.client, err = github.NewEnterpriseClient(baseURL, baseURL, client2)
+		if err != nil {
+			return err
+		}
+	} else {
+		s.client = github.NewClient(client2)
+	}
+
+	return nil
+}
+
+// ListEntries implements the Source interface.
+func (s *GithubReleaseSource) ListEntries(directoryPath string) ([]FileSpec, *ListEntriesError) {
+	return nil, ErrListEntriesNotSupported
+}
+
+// ListAllFiles implements the Source interface.
+func (s *GithubReleaseSource) ListAllFiles(out chan<- FileSpec) *ListEntriesError {
+	releases, err := s.getReleases()
+	if err != nil {
+		return &ListEntriesError{
+			Location: s.url.String(),
+			Message:  "could not list releases",
+			Inner:    err,
+		}
+	}
+
+	var assets []*github.ReleaseAsset
+	for _, r := range releases {
+		if !s.IncludeDraft && r.GetDraft() {
+			continue
+		}
+		if !s.IncludePrerelease && r.GetPrerelease() {
+			continue
+		}
+		if s.TagNamePattern != "" && !s.tagNameRx.MatchString(r.GetTagName()) {
+			continue
+		}
+
+		aL, err := s.getReleaseAssets(r.GetID())
+		if err != nil {
+			return &ListEntriesError{
+				Location: s.url.String(),
+				Message:  "could not list release assets",
+				Inner:    err,
+			}
+		}
+		assets = append(assets, aL...)
+	}
+
+	for _, a := range assets {
+		if a.BrowserDownloadURL != nil {
+			fs := FileSpec{
+				// We need to remove the base URL from the path.
+				Path: strings.TrimPrefix(*a.BrowserDownloadURL, s.url.String()),
+				// Technically, DownloadPath is supposed to be a file path but since we
+				// only need asset ID to download the asset in GetFile() therefore we can
+				// simply use the asset ID here instead.
+				DownloadPath: strconv.FormatInt(a.GetID(), 10),
+			}
+			if a.UpdatedAt != nil {
+				fs.LastModified = &a.UpdatedAt.Time
+			}
+			out <- fs
+		}
+	}
+
+	return nil
+}
+
+// GetFile implements the Source interface.
+func (s *GithubReleaseSource) GetFile(path string, requestHeaders schwift.ObjectHeaders) (io.ReadCloser, FileState, error) {
+	u := fmt.Sprintf("repos/%s/%s/releases/assets/%s", s.owner, s.repo, path)
+	req, err := s.client.NewRequest(http.MethodGet, u, nil)
+	if err != nil {
+		return nil, FileState{}, fmt.Errorf("skipping: could not create URL for %s: %s", u, err.Error())
+	}
+
+	for key, val := range requestHeaders.Headers {
+		req.Header.Set(key, val)
+	}
+	req.Header.Set("User-Agent", "swift-http-import/"+util.Version)
+	req.Header.Set("Authorization", fmt.Sprintf("token %s", s.Token))
+	req.Header.Set("Accept", "application/octet-stream")
+	resp, err := s.httpClient.Do(req)
```

Comment:
Would it also work to just use the default client instead?
```suggestion
	resp, err := http.DefaultClient.Do(req)
```
There is a signifiant difference between `http.DefaultClient` and `http.Client{}` since the former has some sane defaults like following proxies (see also `pkg/util/hacks.go` where we explicitly adjust the DefaultClient to be proxy-friendly).

Reply:
In case I'm not mistaken `http.DefaultClient == http.Client{}` because that is how it's defined [in stdlib](https://cs.opensource.google/go/go/+/master:src/net/http/client.go;l=109).

And [Go doc](https://cs.opensource.google/go/go/+/master:src/net/http/client.go;l=29-30) for `http.Client` says that:

> A Client is an HTTP client. Its zero value (DefaultClient) is a usable client that uses DefaultTransport.

But I forgot about about `http.DefaultClient` behavior modification in `pkg/util/hacks.go` so you're right, we should use `http.DefaultClient` to take advantage of the modified default client for GitHub releases too.